### PR TITLE
Create Server

### PR DESCRIPTION
**Pull request by kulvirs**
Apologies for the large amount of code in this PR! 
I got a bit carried away and decided to actually implement the algorithm for how the server performs updates. I'll try to keep things more modular next time and not bundle so much stuff into one PR.  
Please let me know what suggestions/comments you have.

### Overview
This PR introduces three new packages:
* `objects`
  * contains the JSON objects `File`, `Change`, and `Revision` which are used to communicate information in the requests/responses between the server and client
* `server`
  * contains the server API which has three endpoints for creating files, getting file content, and posting updates to files
  * also contains logic to apply updates on top of previous revisions
* `database`
  * contains an interface which defines all the functions the server needs to interact with the database
  * contains a mock database which implements all the interface functions for unit testing
  * this will eventually contain a file that actually implements the database functions once @NikitaShymberg is done with his database portion

#### Single User Use Case: 
- [x] create a document with an `HTTP POST` request
 - [x] retrieve the latest version of a document with an `HTTP GET` request
 - [x] send updates for a specific document with an `HTTP POST` request
- [x] confirm that the changes took effect by repeating the first step

#### Multi User Use Case:
- [x] retrieve the latest version of a document **R1** with an `HTTP GET` request
- [x] send an update **U1** with an `HTTP POST` request, identifying **R1** as the working copy
    - the latest version of the document on the server should now be **R2**
- [x] send another update **U2** with an `HTTP POST` request, *also identifying **R1** as the working copy*
- [ ] confirm that the server applied **U2** on top of **R2**, even though we sent the update while working on **R1**

Note: The last box is not checked because of some issues that I discovered with our current method of applying updates. A new PR will be made when those issues are (hopefully) fixed.

#### Package Requirements
The `mux` package is required to route requests:
```shell
$ go get github.com/gorilla/mux
```

#### Unit Tests (Go)
For any of the three packages `objects`, `database`, or `server`, perform the following steps to run the unit tests:
```shell
$ cd <package>
$ go test
```
* *NOTE*: use `-v` option for verbose test output

#### API Tests (Postman)
To test the actual server endpoints, I wrote some test scripts in Postman, they can be found at `server/server_tests.postman_collection.json`

To run these tests perform the following steps:
1. Get [Postman](https://www.getpostman.com/apps)
2. Open a terminal and run the server:
```shell
$ cd server
$ go build
$ server
```
3. Open Postman and click `Import`
4. Import the file `server/server_tests.postman_collection.json`
5. Click the arrow next to the `server_tests` folder and then click `Run`
6. In the window that comes up, click `Run server_tests`

### References
- Golang Server API [tutorial](https://www.codementor.io/codehakase/building-a-restful-api-with-golang-a6yivzqdo)

### CODE REVIEW

File: objects/change.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+func NewChange(changeType string, position int, value string) Change {
+	return Change{changeType, position, value}
+}
+
+// Equals defines what makes two Change objects equal.
+func (change *Change) Equals(change2 Change) bool {
+	if change.Type == change2.Type && change.Position == change2.Position && change.Value == change2.Value {
+		return true
+	}
+	return false
```
Comment (okjuan) [inline objects/change.go]: NIT: We could just return the bool value of the if condition

File: server/server.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func main() {
+	router := mux.NewRouter()
+	router.HandleFunc("/file", CreateFile).Methods("POST")
+	router.HandleFunc("/file/{id}", GetFile).Methods("GET")
+	router.HandleFunc("/file/{id}", PostUpdates).Methods("POST")
```
Comment (okjuan) [inline server/server.go]: NIT: resource names are typically plural (`files` instead of `file`)

File: server/server.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+		log.Println("POST request to /file was missing required field(s) or poorly formed.")
+		if err != nil {
+			log.Println(err)
+		}
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	file.ID = database.CreateEmptyFile(file.Name, file.User)
+	file.RevisionNumber = 1
```
Comment (okjuan) [inline server/server.go]: Maybe we could get the revision number as a 2nd return value from the call to `database.CreateEmptyFile`? I assume the database would also be setting it to 1, but that way we have a single source of truth.

File: server/server.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+		// Invalid ID.
+		log.Printf("GET request to /file/%s contained an invalid file ID.\n", params["id"])
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	file, err := database.GetFileContent(id)
+	if err != nil {
+		// Invalid ID.
+		log.Printf("GET request to /file/%s contained an invalid file ID.\n", params["id"])
```
Comment (okjuan) [inline server/server.go]: NIT: Might be useful to make this statement log a different message from the one above (for debugging purposes, for example).

File: server/server.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+			log.Println(err)
+		}
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	file, err := database.GetFileContent(revision.ID)
+	if err != nil {
+		// Invalid ID.
+		log.Printf("POST request to /file/%s contained an invalid file ID.\n", params["id"])
```
Comment (okjuan) [inline server/server.go]: NIT: Same thing here about having a different log statement, maybe something like:
```suggestion
		log.Printf("Could not retrieve file for given ID: %s.\n", params["id"])
```

File: server/server.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+	log.Printf("\"%s\" made updates to File ID %d\n", revision.User, revision.ID)
+	err = ApplyUpdate(revision, file, &database)
+	if err != nil {
+		// Updates were not applied.
+		log.Println("Updates were not applied:", err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
```
Comment (okjuan) [inline server/server.go:102]: Ahh sorry, I originally thought we could avoid returning the new file here, but [apparently](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) that's the norm. So if you still want to return the file contents in the response body, I'm onboard! 🚢

File: server/update.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+			newChange.Position = newDel.Pos
+		case change.Type == "delete" && newChange.Type == "insert":
+			// insertion on deletion
+			newIns := opTransform.TransformInsOnDel(newChange.ChangeToIns(), change.ChangeToDel())
+			newChange.Position = newIns.Pos
+		case change.Type == "delete" && newChange.Type == "delete":
+			// deletion on deletion
+			newDel, err := opTransform.TransformDeletions(newChange.ChangeToDel(), change.ChangeToDel())
+			if err != nil {
+				return newChange, err
```
Comment (okjuan) [inline server/update.go]: NIT: maybe we could return `nil` instead of `newChange` in the error case.

File: server/update.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+			}
+			newChange.Position = newDel.Pos
```
Comment (okjuan) [inline server/update.go]: A `default` case like the one in `ApplyChange` would be nice :-)

File: server/update.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+		}
+		return (fileContent[:change.Position] + change.Value + fileContent[change.Position:]), nil
+	case "delete":
+		if change.Position < 0 || change.Position >= len(fileContent) {
+			// Index out of range.
+			return fileContent, errors.New("index out of range")
+		}
+		return (fileContent[:change.Position] + fileContent[change.Position+1:]), nil
+	default:
+		return fileContent, errors.New("invalid change type")
```
Comment (okjuan) [inline server/update.go:93]: Might be worth adding the index to all these error messages for debugging!

File: objects/change.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+// Change represents each individual change (insertion/deletion).
+type Change struct {
+	// The type of the change (insert or delete).
+	Type string `json:"type"`
+	// The index the change is being made at.
+	Position int `json:"position"`
+	// The value of the character being inserted or deleted.
+	Value string `json:"value"`
+}
```
Comment (okjuan) [inline objects/change.go:17]: I think we should consider somehow merging the `Change` and `Insertion`/`Deletion` objects. I think the fact that we need the `ChangeToDel` and `ChangeToIns` indicates that having a separate `Change` object is possibly redundant.

Perhaps the `Change` object can consist of an `Operation`, which would be an interface that both `Insertion` and `Deletion` implement. Alternatively, we could just completely remove the `Change` object and treat all operations through an `Operation` interface.

**Review by okjuan — CHANGES_REQUESTED** [review]
This is really great, you did an awesome job! I'm stoked about where this project is going 😄 

Most of my feedback consists of small suggestions; the only major idea I have is that we collapse the `Change` object with the `Insertion`/`Deletion` operations. Let me know what you think!

The only files I did not take a close look at are `revision.go` and `file.go`. I think @NikitaShymberg 's opinion on those would be useful, since he's working on the database.

**Review by NikitaShymberg — APPROVED** [review]

File: objects/change.go (hunk 10 above)
Comment (NikitaShymberg) [reply objects/change.go:17]: Agreed - I was picturing it as something as simple as a `nil` Value being a deletion and everything else is an insertion.

**Comment by okjuan** [comment]
@kulvirs Oh yeah, one more thing! Let's update the path from `github.com/jcgallegdup` to `github.com/Dabblr`? 😄

File: objects/change.go (hunk 1 above)
Comment (kulvirs) [reply objects/change.go]: Changed to return bool value of the if condition.

File: server/server.go (hunk 2 above)
Comment (kulvirs) [reply server/server.go]: Changed endpoints to use `files` instead of `file`.

File: server/server.go (hunk 3 above)
Comment (kulvirs) [reply server/server.go]: Good catch, changed the function `database.CreateEmptyFile` to return a file id and revision number.

File: server/server.go (hunk 4 above)
Comment (kulvirs) [reply server/server.go]: Changed invalid ID log messages to be more descriptive.

File: server/server.go (hunk 5 above)
Comment (kulvirs) [reply server/server.go]: Similar to above, changed log messages for invalid id to be more descriptive.

File: server/server.go (hunk 6 above)
Comment (kulvirs) [reply server/server.go:102]: Hmm...wait so do you want to change it, bit confused what you're asking. Currently the `PostUpdates` endpoint just returns 200 if the request is successful, and 400 if it is not. No body is returned, I thought we were going to assume that the client would call `GET` after `POST`?   
If it is proper convention to return the content with the `POST` request, I can change it to do that.

File: server/update.go (hunk 7 above)
Comment (kulvirs) [reply server/update.go]: I wanted to do this too but you can't use a `nil` value for a struct. So we're kind of stuck with this.

File: server/update.go (hunk 8 above)
Comment (kulvirs) [reply server/update.go]: Added a default case.

File: server/update.go (hunk 9 above)
Comment (kulvirs) [reply server/update.go:93]: Added index to error messages.

File: objects/change.go (hunk 10 above)
Comment (kulvirs) [reply objects/change.go:17]: As per our discussion last Thursday, we are going to leave this as it is for now. I've created an [issue](https://github.com/Dabblr/Concurrent-Document-Editor/issues/4) for it so we can remember to come back to it if we figure out a clean way to do this.

**Comment by kulvirs** [comment]
Updated path from `github.com/jcgallegdup` to `github.com/Dabblr`. Note that you may have to change the name of that folder manually in your own local repo for the imports to work.

**Review by okjuan — APPROVED** [review]
As long as the tests still pass, this looks great to me!

**Pull request merged**